### PR TITLE
[stdlib] Consolidate the standard library in its own directory.

### PR DESCRIPTION
**Pull request by ejgallego**
Not sure this is a good idea, however, it brings quite a bit of
uniformity and will allow to build the stdlib with Dune as a regular
library.

Alternatives are a bit cumbersome:

- a fine-grained dependency chain is a bit tricky, and high maintenance,
- a merge mode is not justified enough I think.

I disabled the whitespace fixer as it would break git's renaming
algorithm.

### CODE REVIEW

**Comment by ejgallego** [comment]
cc #6137

**Comment by gares** [comment]
How is stdlib2 accessing, say, lia in this scenario?

**Comment by Zimmi48** [comment]
I would have opened this as wip to avoid spamming every code owner until we actually take a decision. But well, it's too late to make such a recommendation.

**Review by maximedenes — COMMENTED** [review]
I thought you wanted to move `plugins` and `theories` to `stdlib/plugins` and `stdlib/theories`.

**Comment by Zimmi48** [comment]
In fact, I'd recommend closing this PR and opening a CEP instead. There is nothing interesting regarding implementation here, it's all about design.

File: Makefile.doc [hunk 1]
```diff
@@ -90,7 +90,7 @@ doc-pdf:\
 doc-ps:\
   doc/stdlib/Library.ps
 
-stdlib: \
+light-stdlib: \
```
Comment (SkySkimmer) [inline Makefile.doc]: Why `light`? Shouldn't it be `stdlib-doc` or some such instead?

File: tools/coq_dune.ml [hunk 2]
```diff
@@ -214,7 +214,7 @@ let record_dune d ff =
   if Sys.file_exists sd && Sys.is_directory sd then
     let out = open_out (bpath [sd;"dune"]) in
     let fmt = formatter_of_out_channel out in
-    if List.nth d 0 = "plugins" then
+    if List.nth d 0 = "stdlib" && Sys.file_exists (bpath [sd;"plugin_base.dune"]) then
       fprintf fmt "(include plugin_base.dune)@\n";
```
Comment (SkySkimmer) [inline tools/coq_dune.ml]: When does plugin_base exist but we don't want to include it? ie why the list.nth part of the test?

**Comment by ejgallego** [comment]
> I thought you wanted to move `plugins` and `theories` to `stdlib/plugins` and `stdlib/theories`.

Any arrangement is possible, but that interacts a bit worse w.r.t. paths, well, it depends if you put the files in `plugins/*/.v` in theories or not.

So if you have `plugins/foo/Bar.v` it has to be grafted into the Coq hierarchy, that seems a bit ugly.

File: tools/coq_dune.ml (hunk 2 above)
Comment (ejgallego) [reply tools/coq_dune.ml]: That's just a hack, `coq_dune` should go away next week anyways so I wouldn't worry too much about this.

File: Makefile.doc (hunk 1 above)
Comment (ejgallego) [reply Makefile.doc]: Maybe, here I had to do this a very weird things were happening with the makefile. I am not sure things are still working OK.

**Comment by ejgallego** [comment]
> So if you have `plugins/foo/Bar.v` it has to be grafted into the Coq hierarchy, that seems a bit ugly.

Like this you can work just with `-R stdlib Coq`. Seems more consistent.

The ML-side `-I` paths are hopefully going away for 8.11 as we use META information to locate plugins [and correctly load dependencies!]

> In fact, I'd recommend closing this PR and opening a CEP instead. There is nothing interesting regarding implementation here, it's all about design.

You are right about the draft status, sorry; however there is code to review here, in particular to understand the impact this has on some subsystems and CI, so I am keeping the PR.

**Comment by gares** [comment]
I have the impression that if you don't `git mv theories/* plugins/* stdlib/*` but you simply add an directory called `stdlib` and add a dune rule executing `ln -s theories/* plugins/* stdlib` we get a situation where you can build `stdlib` with dune without moving any file but just adding instead of replacing paths in coqinit. You would not need to change the Makefiles at all. When Makefiles are removed, you prune the two old paths from coqinit, and do the `git mv`.

**Comment by cpitclaudel** [comment]
I have no opinion on the actual content of this PR, but I'll not that until now region history (`git log -L` or `M-x vc-region-history`; see https://stackoverflow.com/questions/23877281/see-history-of-region-in-git) on Coq source files has worked nicely (often going back to the CVS import in 2000). Giant renames tend to break region history, which isn't great.

**Comment by SkySkimmer** [comment]
>I have no opinion on the actual content of this PR, but I'll note that until now region history (git log -L or M-x vc-region-history; see https://stackoverflow.com/questions/23877281/see-history-of-region-in-git) on Coq source files has worked nicely (often going back to the CVS import in 2000). Giant renames tend to break region history, which isn't great.

It works fine for me. For instance selecting https://github.com/coq/coq/blob/ae4239d6a5f4afcd9b7321dba790ffd4a64994a1/interp/declare.ml#L108-L115 and running vc-region-history shows 3fcf0930874d7200f2503ac7084b1d6669d59540 908dcd613b12645f3b62bf44c2696b80a0b16940 etc, note that at the time of the second one declare.ml was in library/

**Comment by cpitclaudel** [comment]
> It works fine for me.

Neat! I'm not sure what the criteria are to keep it working. But my fears seem overblown, then.

**Comment by ejgallego** [comment]
> I have the impression that if you don't `git mv theories/* plugins/* stdlib/` but you simply add a directory called `stdlib` and add a dune rule executing `ln -s theories/* plugins/* stdlib` we get a situation where you can build `stdlib` with dune without moving any file by just adding (instead of replacing) paths in coqinit, etc. You would not need to change the Makefiles at all. When Makefiles are removed, you prune the two old paths from coqinit, and do the `git mv`.

Sure, we can do that; Dune doesn't need in fact any change to build the stdlib, the setup is just a bit more messy; indeed this is more a matter of organization maybe than a matter of the build system / stdlib2 / modularity. So I dunno, maybe it is a good idea, maybe not. Indeed if we are going to do the `mv` why to delay it.

> note that until now region history (git log -L 

Modern git toolchain should not have a problems with these renamings; Coq indeed already has a few of those. I was careful in fact to test beforehand.

Something that is breaking renaming tracking in our setups tho is the auto-whitespace cleanup; people moving things around should be careful with it.

Regarding the build failure, I am at a total loss over what make weird interaction is triggering this time. Super weird.

**Comment by SkySkimmer** [comment]
>Regarding the build failure, I am at a total loss over what make weird interaction is triggering this time. Super weird.

build:base uses the `stdlib` target whose meaning has changed.

**Comment by ejgallego** [comment]
> build:base uses the `stdlib` target whose meaning has changed.

ahh, thanks! I looked at the wrong place.

**Comment by Zimmi48** [comment]
> Indeed if we are going to do the `mv` why to delay it.

The only good reason I see to delay it is if we have still unclear design decisions. E.g. do we want historic plugins be in the `Coq` namespace and the stdlib package or in their own namespace / package? It could make sense to put them in their own namespace / package so that e.g. stdlib2 can depend on them more easily. OTOH, most of these plugins actually depend on stdlib.
Do we actually want to preserve the `Coq` namespace or do we want to call it `Stdlib`? I guess this would be breaking things for no good reason though.

**Comment by Zimmi48** [comment]
For instance, what about a single `packages/` directory that would contain prelude, datatypes, arith, reals, ltac, ltac2, ssreflect, micromega, unicoq, etc., each of them being a separate package?

**Comment by ejgallego** [comment]
> Do we actually want to preserve the `Coq` namespace or do we want to call it `Stdlib`? I guess this would be breaking things for no good reason though.

Indeed IMHO it is out of the question to change this, unless in this first "iteration". Maybe when we have a "2.0" package format we can allow people to migrate to a better namespace.

> For instance, what about a single `packages/` directory that would contain prelude, datatypes, arith, reals, ltac, ltac2, ssreflect, micromega, unicoq, etc., each of them being a separate package?

Actually that would be cool, however the inter-dependencies among packages are, as of today, quite tricky and not always expressible in the form of a DAG!!

So uh! Modularizing `stdlib` will take some effort indeed, so IMHO it is better to see it in as a single package as of today.

**Comment by ejgallego** [comment]
> not always expressible in the form of a DAG!!

This is an hypotheses, it is been quite a long time I did an analysis of the full dep graph, but I recall I stopped with this impression.

**Review by JasonGross — CHANGES_REQUESTED** [review]
Please also adjust `TimeFileMaker.py` as follows to maintain display of timing tables:
```diff
-STRIP_REG = re.compile('^(coq/|contrib/|)(?:theories/|src/)?')
+STRIP_REG = re.compile('^(coq/|contrib/|)(?:theories/|src/|stdlib/)?')
```

**Comment by ejgallego** [comment]
> Please also adjust `TimeFileMaker.py` as follows to maintain display of timing tables:

Done, thanks for the pointer.

**Comment by ejgallego** [comment]
Hi all, first sorry for the review spam , I'll be happy to close this and reopen as a draft if that's too annoying.

Second, I'd be great if we could have some reviews / comments on the current form of this PR, as this is the base branch I'm using for implementing the bootstrap-free version of the Dune build.

Other options are possible, but IMHO this makes for the simpler setup, one single `dune` file in `stdlib/theories` and one regular OCaml plugin file in each of the ML `stdlib/plugins` directories.

**Comment by gares** [comment]
> Second, I'd be great if we could have some reviews / comments on the current form of this PR, as this is the base branch I'm using for implementing the bootstrap-free version of the Dune build.

This PR is unreviewable, I think it is best if you tell us precisely what and why.

**Comment by ejgallego** [comment]
> This PR is unreviewable, I think it is best if you tell us precisely what and why.

The PR does move `plugins` and `theories` into `stdlib`, it also separates the ml part from the `.v` part; the idea is that then `stdlib` is self-contained into a single directory.

The separation of the ml parts from the v parts is due to the way `(include_subdirs ...)` work. If all the `.v` parts are inside a single directory hierarchy then you can just add a `dune` file at the root and use `(include_subdirs qualified)`; if you add `.ml` files there you have to mark manually all the dirs with OCaml code as not being qualified as OCaml does not support that yet. That is a bit tedious.

**Comment by maximedenes** [comment]
Let's discuss this during the next Coq call, I'd like to understand better the implications on the development of alternative standard libraries (stdlib2) which would like to reuse some of these plugins.

**Comment by ejgallego** [comment]
> Let's discuss this during the next Coq call,

Sure, I already added the item.

> I'd like to understand better the implications on the development of alternative standard libraries (stdlib2) which would like to reuse some of these plugins.

I don't think this PR in particular does change the status quo, the build is the same for now; this just moves the files around. The `dune` build using this infra does indeed build the `Coq` prefix as a single entity, as done in the Makefile.

For the ML part there should be no change or problem, plugins are just packed OCaml libs.

The `.v` part is a bit different, as of today plugins live in `Coq.plugin.Vfile` and they depend on the prelude already, there are indeed a few choices to make here to make the plugins more independent but as far as I can see changes would be required.

If you can provide some example I could try to do some dependency analysis for the call so we understand the situation better.

File: .github/CODEOWNERS [hunk 3]
```diff
@@ -172,13 +172,72 @@
 
 ########## Standard library ##########
 
+<<<<<<< HEAD
```
Comment (gares) [inline .github/CODEOWNERS]: there is a conflict here

**Comment by maximedenes** [comment]
Could we get the list of `git mv` commands used to produce this PR? It would make it easier (possible) to review it.

**Review by maximedenes — CHANGES_REQUESTED** [review]
I would prefer the ML part of plugins to stay in `plugins`, not `stdlib/plugins`. It would make it easier to develop other libraries which do not depend on stdlib, but still want to use the ML code of plugins.

**Comment by ejgallego** [comment]
I am closing this in favor of #11529

**Pull request closed without merging**